Configurable priorities

TamsenSpelchak opened the pull request:
Allow `tool_destinations.yml` to contain as many arbitrary priorities as are listed in the `default_destination` section.

Also, have `dynamic_tool_destination.py` declare `tool_destinations.yml` as invalid if it references destinations that don't exist in `job_conf.xml`, and suggest corrections for possible typos in destination and priority names.

Commit:
Allowing for configurable priorities.

Any priority listed in the global 'default desinations' section in the tool destinations config is considered a valid priority. All other priorities are invalid.
This allows for the configuration and usage of an arbitrary number of priorities

Commit:
checking for valid destinations in tool destination config

dynamic_tool_destination.py now checks that any job destination referenced
exists in job_conf.xml by matching the name of the destination in
tool_destinations.yml to a destination id in job_conf.xml

Commit:
some code cleanup and changing default priority on tools

The default priority for tools if not specified in the tool destination config is now whatever the middle priority is in the list of global default destinations

Commit:
Fixing some error messages

Commit:
adhering (closer) to pep8

Commit:
fixing bug

Commit:
Fix typo in validated config; test all destinations for validity

Previously, the validate_config method would return a config with a
'default' section instead of 'destination_default'. This caused
some unit tests to error out and raise a JobMappingException. This
has been fixed.

Previously, only destinations that went through the validate_rules
method would be checked to ensure existence. Now, destinations in
top-level 'default_destination' and tool-level
'default_destination' are also checked.

Commit:
Fix error with validated configs

Commit:
Change priority and destination lists to sets

By using sets instead of lists for the valid priority and destination,
the checks for item uniqueness in these datatypes is now handled
inherently, with little detriment.

The only drawback to this is that the system that chooses the default
priority for each tool now simply chooses the first element from
iter(), rather than the one in the middle of the list. This isn't a
huge deal, as that was no more than a guess at what a mid-level
priority would be, and would even cause errors when the chosen
priority wasn't in the tool's default destination section.

Using sets insetad still result in the same issue, so this will need
to be fixed soon, as it prevents certain tool destination configs
from working properly.

Commit:
Only call get_valid_destinations_from_config once

Commit:
Implement hack to fix tools not having destinations for default priority

Commit:
Make users adhere to same priorities as tools, better handle tools without default priority destination, stop destination and priority sets from retaining values between test cases, updating tests

Previously, user's had their own sets of priorities (low, med, high)
that did not match the new system with configurable priorities. As
mapping these user priorities onto tool priorities would be impossible
without extra config sections, or hardcoding of things that map to one
another, the user priorities are now on the same system. IE: all
priorities declared in the user sections much match priorities in the
"default_destination" section.

Dynamic_tool_destination also now uses a less-hacky method of handling
tools that don't have a default destination for the default priority
(which is still unfortunately arbitrarily chosen; this presents a
rather large issue that will need to be addressed in the future.)
Essentially, the global default destination for the default priority
is used if there isn't one defined in the tool's defaults or its rules.

There was also a bug previously that caused dynamic_tool_destination's
priority_list and and destination_list (formerly valid_destinations)
to retain values in between test cases, causes the test results to
differ from the expected results. This bug has been fixed.

Finally, the test_dynamic_tool_destination program has been updated
to reflect the exptected outcome of test cases with the new program's
expected functionality in mind. Note that this does not mean the tests
are testing what they should be -- they simply don't fail anymore for
expected behaviour.
There is test that still fails -- and it deals with the fact that
there no longer is a default priority that all tools need to adhere
to. Because default priorities are determined now in an arbitrary way
(this shouldn't be permanent behaviour), a destination other than the
expected one may be chosen for a given tool that does not implement
that (arbitrarily chosen) default priority.

Commit:
Ignore backup files

I append ".bak" to copies of new files that I'm working on as a
quick and dirty backup method. Updated the .gitignore file to
ignore these

Commit:
Fix error messages, add tests, miscellaneous fix

A few error messages have been altered to be a bit clearer, and a
bug was fixed that stopped an error from being fired.

Tests were also added to test some of the new functionality of
dynamic_tool_destination. The added tests:

 - test for an error being thrown when a priority in priority_list
   does not appear in a tool's default_destination section (this
   may not be final behaviour)

 - test to ensure that errors are fired if a destination is
   referenced that is not in the destinations section of the
   job_conf.xml

Commit:
Add more tests, make error messages more specific

Three new tests were added to ensure that errors are thrown if a
destination is used as a straight string value (without a priority)
and that destination does not exist in the job_conf.xml

Also updated a few error messages to explicitely mention that the
destination is not in job_conf.xml, hopefully ensuring users
who get the message know what to do to fix the error.

Commit:
Alter error messages, add new tests

A few error messages were altered to start with lowercase letters.
Why? Don't ask me, that was yesterday. Expect all the error
messages to be altered to start with capitals (as I believe they
should) soon.

Two new tests were added, ensuring that errors are fired when
priorities are referenced that don't appear in the config's global
default_destinaion section.

Commit:
Make all error messages start with capitals, and adjust the tests to expect this

Commit:
Allow job config xml files to be passed as an argument

job configuration xml files can now be passed as arguments to the
application with the -j  or the --job-config options. To
facilitate this change, the parse_yaml constructor has been altered
to accept a new parameter, job_conf_path, the path to the job
configuration.

For simplicity, get_destination_list_from_job_config is now called
from parse yaml instead of validate_config. This simply reduces the
amount of times the job_conf path needs to be passed between
functions.

Commit:
Tests include their own job_conf.xml, fix bug when specifying job_conf.xml paths, code cleanup

The built-in dynamic_tool_destination tests now use a job_conf.xml
included alongside the testing tool destination yaml files. This
means that DTD can finally be tested in an automated fashion -- no
more manually putting a job_conf.xml into the config folder to make
things work.

This means that map_tool_to_destination now has an additional
parameter, job_conf_path -- the path to the job_config.xml file.

Also fixed is a bug that would add the Galaxy config folder path to
custom job_conf.xml paths.

Finally, the tests that were not relevant to the new version of DTD
that were commented out are now removed altogether.

Commit:
Clean up comments

Commit:
Follow pep8 (mostly)

Commit:
Adhering (better) to Galaxy style guidelines

Commit:
Adjust doc comments to 79 character max line length

Commit:
Shorten logical lines over 79 characters where it makes sense

Commit:
Trim test configs, add new tests

The configurations used by the DTD unit tests have been trimmed,
stripping out parts that do not contribute to test results.

Two new tests were also added, both testing to ensure that a tool
is reported to not have any default destinations *or* rules if its
only default destination is invalid, and it has no rules.

Commit:
Fix tests that didn't run.

Previously, two tests would not run because I forgot to name them
to something that wasn't a copy/paste from another test.

I renamed them so they actually run.

Commit:
Add ability to set default priority, update tests, small fixes

This is a relatively large commit -- a new section in
tool_destinations.yml is now required: default_priority, which sets
which priority to be used as default for all tools, replacing the
way 'med' was previously assumed to be default, back when
priorities were not configurable.

As a fallback to ensure compatibility with the old style of config,
if no default_priority is specified in the config, but 'med' exists
as a valid priority, the priority will be set to 'med'.

Additionally, tests were updated to expect this new behaviour.

A collection of small bug fixes were also added, such as:
 - not adding anything to the list of priorities when the global
   default_destination is only a string,
 - Making error messages clearer
 - fixing some indentation

Commit:
Ensure default_destination for each rule contains a destination for default_priority

Commit:
Revert previous commit

It doesn't make sense to have the config validation fail just
because a tool doesn't have a destination for the default priority:
the destination should just go to the global default destination,
as the user would likely expect to happen.

Commit:
Fix error message order, add test

Fixed some error message logic. Previously, a tool_destinations.yml config
with an empty priority key in global 'default_destination' would throw two
errors in a strange order; one about no valid default priority destinations,
and another about no default destination. This has been changed so with the
same config there is now just a 'No default destinations in config' message.

Also added a test that confirms this behaviour

Commit:
Add new tests, add new category to valid keys check

Added three new tests, testing cases where all destinations are str,
where the default_destination is a str but a tool
default_destination is a dict, and where the default_destination is
a str but a tool rule destination is a dict.

Also fixed a bug that was caught with these new tests --
default_priority was not actually in the list of valid config
categories so using the default_priority key would result in errors.
Whoops.

Commit:
Reorganize test code

Commit:
Make new tests actually run

Commit:
Merge branch 'dev' of https://github.com/phac-nml/galaxy into configurable_priorities__default_priorities

Commit:
Fix default priority assignment order

Moved where default_priority is actually generated in the
map_tool_to_destination function. Now it calculates and assigns
default_priority to priority *before* the job/workflow parameter-
specified priority is assigned.

Commit:
Add comments to describe some of the new functionality

Commit:
Adhere to Galaxy Python standards

Commit:
Fix test that didn't run

Commit:
Begin adding system to suggest typo corrections

Implemented a function to check for edit distance that I found
online, and wrote a function that compares a given string with all
entries in a set and returns the one with fewest edits.

Added code to validate_destination in RuleValidator to suggest
corrections for possibly mispelled priorities, but haven't tested
it yet.

Commit:
Finish implementing typo correction suggestions

Code has been added in all locations in the code that check values
against the list of valid values, and checks if there are similar
valid values in the list if the value in the config is invalid.

If there is a similar value (edit distance 3 or closer, currently,)
output will be printed pointing out the similar, valid value.

Commit:
Fix max_edit_dist not being set or used

Commit:
Change max_edit_dist to 2

Commit:
Start adding tests

Added two tests to check if correct destinations were recommended
when 'waffles_low' was mispelled in default destinations sections.

Commit:
Add more tests

Test that correct suggestions occur in all cases where values that
don't match the ones in priority_list or destination_list are used.

Commit:
Add case insensitive comparison to get_typo_correction, new test to test this.

get_typo_correction has been updated to prefer suggesting strings
that are identical after .lower() is run on both.

A new test has also been added to help confirm this new
functionality.

Commit:
Remove old comment

Commit:
Merge branch 'dev' into configurable_priorities

Commit:
Merge branch 'dev' of https://github.com/phac-nml/galaxy into configurable_priorities__default_priorities__guess_typos

Takadonet commented on test/unit/jobs/dynamic_tool_destination/ymltests.py:
```diff
@@ -945,3 +945,167 @@
+ivYMLTest149 = '''
+    default_destination:
+      priority:
+        med: waffles_low
```
get rid if waffles name please.

Takadonet commented on lib/galaxy/jobs/dynamic_tool_destination.py:
```diff
@@ -1446,11 +1598,21 @@ def map_tool_to_destination(
 
     else:
-        valid_config = parse_yaml(path="/config/tool_destinations.yml", return_bool=True)
+        valid_config = parse_yaml(path="/config/tool_destinations.yml",
```
also check for other version of the job_conf.xml.sample_advanced and/or job_conf.xml.sample_basic

Takadonet commented on lib/galaxy/jobs/dynamic_tool_destination.py:
```diff
@@ -1226,9 +1319,21 @@ def map_tool_to_destination(
+    # making default destinations a mandatory field for all tools
+    if len(priority_list) > 0:
+        default_priority = next(iter(priority_list))
+        priority = default_priority  ###FIXME: this is sometimes a value that isn't in the tool's default destinations field
```
fixme?

Commit:
remove references to waffles

Commit:
Merge branch 'configurable_priorities' of https://github.com/phac-nml/galaxy into configurable_priorities

TamsenSpelchak commented:
Accidentally had the wrong branch here at first, whoops. I pushed the correct branch and will now address your comments.

Takadonet commented:
Please update the tool version as well please.

TamsenSpelchak commented:
What would be an appropriate version?

Takadonet commented:
1.1 :)

Commit:
Check for sample job_conf files if the standard one doesn't exist, other miscellaneous fixes

Updated __main__ code and get_destination_list_from_job_config  to
check for the sample job config files if the default one doesn't
exist, and if in verbose mode, notify the user to which one will
be used.

Also changed the version number to 1.1

Also removed references to waffles in the test files

Commit:
Better handling of no job configs, fixing indentation

Now, if no job config is specified and none can be found in the
'config/' folder, rather than crashing trying to access a file that
doesn't exist, the program just keeps the destination_list empty,
resulting in a bunch of "destination not found" errors (as expected,)
but (hopefully) no crashing.

Also fixed weird 2-space indentation that happened in the previous
few commits.

Commit:
Fix nesting error

A simple error in indentation that caused job configs to not be read
properly has been fixed.

TamsenSpelchak commented:
Fixed everything you reviewed so far that needed fixing!

ericenns commented on lib/galaxy/jobs/dynamic_tool_destination.py:
```diff
@@ -973,8 +1124,9 @@ def infinite_defaultdict():
         for category in obj.keys():
-            if not (category == 'verbose' or category == 'tools' or
-                    category == 'default_destination' or category == 'users'):
+            if not (category == 'verbose' or category == 'tools'
```
might be better to check if `category in valid_categories` easier to add a category in the future then.

ericenns commented on lib/galaxy/jobs/dynamic_tool_destination.py:
```diff
@@ -1416,14 +1591,183 @@ def map_tool_to_destination(
     return destination
 
 
+def get_destination_list_from_job_config(job_config_location):
```
Is there a reason why you are not using galaxy's built in job_config parser and using the get_destination method https://github.com/galaxyproject/galaxy/blob/6d94009cb359ef58acd987a0fbdf3f9d024c7c17/lib/galaxy/jobs/__init__.py#L536

TamsenSpelchak replied:
Hehe, because I didn't know it existed. Perhaps I should've done a bit more research. I'll try the built-in parser!

TamsenSpelchak replied:
Would you happen to know if it's possible to grab all destinations using [`get_destinations`](https://github.com/galaxyproject/galaxy/blob/6d94009cb359ef58acd987a0fbdf3f9d024c7c17/lib/galaxy/jobs/__init__.py#L551), like by passing in an empty string or wildcard or something? I guess the other option would be throwing out the `destination_list` completely and checking if destinations are valid on the fly with [`get_destination`](https://github.com/galaxyproject/galaxy/blob/6d94009cb359ef58acd987a0fbdf3f9d024c7c17/lib/galaxy/jobs/__init__.py#L536).

TamsenSpelchak replied:
After some more digging, it looks like [`get_destinations`](https://github.com/galaxyproject/galaxy/blob/6d94009cb359ef58acd987a0fbdf3f9d024c7c17/lib/galaxy/jobs/__init__.py#L551) cannot be run outside the core galaxy code, and I don't see any equivalent functions in [galaxy-lib](https://github.com/galaxyproject/galaxy-lib). Because of this, I believe parsing `job_conf.xml` inside DTD is the best option, as it allows users to validate configurations with destination-checking by running the file directly, in addition to having destination-checking occurring at Galaxy runtime.

ericenns replied:
Ok, so when the destination is running you can use app.job_config.get_destination(destination), but in offline validation mode you do not have access to the app object. One potential solution is to use your parser optionally in offline mode so that if the job_config.xml get's changed in the future your code won't break. And when in online mode use the app object to get_destination. Though I will defer to @Takadonet for his opinion.

Takadonet replied:
Think the best options would be to only use your parser when in offline mode. So if job_conf.xml changes, only validation checking is broken and not 'online' mode.

TamsenSpelchak replied:
Sounds like a solid plan!

Commit:
Create list of valid categories

Upgraded check for valid categories by creaing a valid_category
global variable containing all of the valid categories that are
accepted.

Commit:
When in online mode, use galaxy's get_destination method

TamsenSpelchak commented:
Pushed draft of the online and offline destination-checking. It may still need a bit of cleaning up, but this is the general idea of how it works.

ericenns reviewed (changes requested):
I suggest that you refactor your destination validation into a valid_destination function that takes as input app and destination and return a bool.

In the same vein I would create a suggestion function that takes app as input and whatever else is required and returns none if app is none otherwise a suggestion if one is found.

TamsenSpelchak commented:
Got it. If I'm passing `app` to it should I not be making `app` a global variable anymore?

TamsenSpelchak commented:
Idea: to reduce all of the checking if `app` exists each time to determine if we want to do a correction, have the `validate_destination` function (which will take `app` as a parameter) that does all of that checking, and returns a string if it's invalid and there's a suggestion, false if it's invalid and there's no suggestion, and true if it's valid. Or would it be bad practice to have a function that returns different types depending on outcome?

ericenns commented:
It should be consistent, I guess you could always return a tuple first item
a boolean and second item a suggestion

On Thu, Mar 1, 2018 at 10:12 AM Matthew Spelchak <notifications@github.com>
wrote:

> Idea: to reduce all of the checking if app exists each time to determine
> if we want to do a correction, have the validate_destination function
> (which will take app as a parameter) that does all of that checking, and
> returns a string if it's invalid and there's a suggestion, false if it's
> invalid and there's no suggestion, and true if it's valid. Or would it be
> bad practice to have a function that returns different types depending on
> outcome?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/phac-nml/galaxy/pull/3#issuecomment-369642815>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAeCX2cakeUK9sYwpgbUV9DErvwzhynrks5taB4AgaJpZM4SKwgQ>
> .
>

TamsenSpelchak commented:
I'll do that, then; it seems like the best way to make everything simpler to _and_ reduce lines of code.

Commit:
Create method for verifying destinations

Now, destinations are validated in a function that returns
both the validity of destination, as well as a suggestion
for typo correction if a typo is suspected. This function
is what does the check for if `app` exists now, making
validating destinations use much less lines of code now,
and also making these sections of code much easier to read.

Commit:
pass `app` instead of having it be a global var

TamsenSpelchak commented:
Pushed my latest changes. `app` is no longer a global variable, instead it is passed to the functions that require it. Additionally there is now a function that validates destinations, which handles all of the checks to see if the code is being run in 'online' or 'offline' mode, and validates accordingly.

Commit:
Alter `validate_destination` to also print error messages

Now, `validate_destination` is also passed an error message to
format, and a tuple of values to be be used in the formatted
message. All code that calls validate_destination has been updated
to acknowledge this new functionality.

TamsenSpelchak commented:
I've updated `validate_destination` to also print error messages. I think I can still make things cleaner though, by only having `validate_destination` called once for each part of the config, rather than two calls the way it is now: one in the event of a `dict` destination, and one in the event of a `str` destination.

TamsenSpelchak commented:
I looked back at how I was setting `valid_config` to the boolean result of `validate_destination` in other parts of the code, and I believe that it also causes similar issues there. I'm going to write a quick test to see if I can prove this, and if it's the case, I'll update DTD accordingly.

Commit:
Add New Tests, Fix Bug Uncovered From Test

Added two new tests to test for `valid_config` not being messed with by
previous assignments to it.

This proved that the previous design of setting `valid_config` to the
value of `validate_destination` in `validate_config` caused errant
behaviour when a default destination was invalid but a tool default
destination was valid. This case caused the program to still end up
returning `valid_config` as True even when an invalid default destination
was found.

TamsenSpelchak commented:
Just pushed a new commit that changes the way `valid_config` is assigned in `validate_config`, and fixes a bug that was found with 2 new tests that were also added.

Commit:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into configurable_priorities

Commit:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into dev

Commit:
Merge branch 'dev' into configurable_priorities

Commit:
Fix Formatting

Commit:
Adhere to Linting Rules

Commit:
Try To Pass Linting (Again)

Commit:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into configurable_priorities

Commit:
Adhere to Linting Rules (Again Again)

TamsenSpelchak committed:
Merge Galaxyproject Dev, Remove Modifications on Unrelated Files

TamsenSpelchak committed:
Remove Unrelated Files (Again)

TamsenSpelchak committed:
Reset Unrelated Files to Their State in upstream/dev

TamsenSpelchak committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into configurable_priorities

TamsenSpelchak committed:
Reset One More File to Their State in upstream/dev
